compiler/hash-typecheck/src/traverse/mod.rs

```diff
@@ -2050,10 +2031,49 @@ impl<'gs, 'ls, 'cd, 'src> visitor::AstVisitor for TcVisitor<'gs, 'ls, 'cd, 'src>
 
     fn visit_list_pat(
         &mut self,
-        _ctx: &Self::Ctx,
-        _node: hash_ast::ast::AstNodeRef<hash_ast::ast::ListPat>,
+        ctx: &Self::Ctx,
+        node: hash_ast::ast::AstNodeRef<hash_ast::ast::ListPat>,
     ) -> Result<Self::ListPatRet, Self::Error> {
-        todo!()
+        let walk::ListPat { elements } = walk::walk_list_pat(self, ctx, node)?;
+
+        // We need to collect all of the terms within the inner pattern, but we need
+        // have a special case for `spread patterns` because they will return `[term]`
+        // rather than `term`...
+        let inner_terms = elements
+            .iter()
+            .zip(node.fields.iter())
+            .filter(|(_, node)| !matches!(node.body(), hash_ast::ast::Pat::Spread(_)))
+            .map(|(element, _)| -> TcResult<TermId> { self.typer().get_term_of_pat(*element) })
+            .collect::<TcResult<Vec<_>>>()?;
+
+        let list_term = self.unifier().unify_rt_term_sequence(inner_terms)?;
```

Comment:
Unifying a rt term sequence of types??

Reply:
Admittedly the naming of this function is terrible. The function expects all of the terms provided to be wrapped in `Rt`, since it is getting the `type_of` each term and then performs the unification. it is unclear what the function should be named, any ideas?

Reply:
Oh sorry, I thought you were getting the types of the patterns rather than their values. Nvm, this is ok.

Reply:
Also, is this not right since when the patterns are typed, they are all wrapped in `Rt`.

Reply:
Not sure what you mean by that